packages/playwright-core/src/server/socksClientCertificatesInterceptor.ts

```diff
@@ -0,0 +1,213 @@
+    // Workaround: Rewrite to localhost during tests.
+    this.target = await createSocket(isUnderTest() ? 'localhost' : this.host, this.port);
+    this.target.on('close', () => this.socksProxy._socksProxy.sendSocketEnd({ uid: this.uid }));
+    this.target.on('error', error => this.socksProxy._socksProxy.sendSocketError({ uid: this.uid, error: error.message }));
+    this.socksProxy._socksProxy.socketConnected({
+      uid: this.uid,
+      host: this.target.localAddress!,
+      port: this.target.localPort!,
+    });
+  }
+
+  public onClose() {
+    this.internal?.destroy();
+    this.target.destroy();
+  }
+
+  public onData(data: Buffer) {
+    // HTTP / TLS are client-hello based protocols. This allows us to detect
+    // the protocol on the first package and attach appropriate listeners.
+    if (!this.firstPackageReceived) {
+      this.firstPackageReceived = true;
+      // 0x16 is SSLv3/TLS "handshake" content type: https://en.wikipedia.org/wiki/Transport_Layer_Security#TLS_record
+      this.isTLS = data[0] === 0x16;
+      if (this.isTLS)
+        this._attachTLSListeners();
+      else
+        this.target.on('data', data => this.socksProxy._socksProxy.sendSocketData({ uid: this.uid, data }));
+    }
+    if (this.isTLS)
+      this.internal!.push(data);
+    else
+      this.target.write(data);
+  }
+
+  private _attachTLSListeners() {
+    assert(this.isTLS);
+
+    this.internal = new SocksConnectionDuplex(data => this.socksProxy._socksProxy.sendSocketData({ uid: this.uid, data }));
+    this.internalTLS = new tls.TLSSocket(this.internal, {
+      isServer: true,
+      // openssl req -new -newkey rsa:2048 -days 3650 -nodes -x509 -keyout key.pem -out cert.pem -subj "/CN=localhost"
+      key: fs.readFileSync(path.join(__dirname, '../../bin/socks-certs/key.pem')),
+      cert: fs.readFileSync(path.join(__dirname, '../../bin/socks-certs/cert.pem')),
+    });
+    this.internalTLS.on('close', () => this.socksProxy._socksProxy.sendSocketEnd({ uid: this.uid }));
+
+    const targetTLS = tls.connect({
+      socket: this.target,
+      rejectUnauthorized: this.socksProxy.contextOptions.ignoreHTTPSErrors === true ? false : true,
+      ...clientCertificatesToTLSOptions(this.socksProxy.contextOptions.clientCertificates, `https://${this.host}:${this.port}/`),
+    });
+
+    this.internalTLS.pipe(targetTLS);
+    targetTLS.pipe(this.internalTLS);
+
+    // Handle close and errors
+    const closeBothSockets = () => {
+      this.internalTLS?.end();
+      targetTLS.end();
+    };
+
+    this.internalTLS.on('end', () => closeBothSockets());
+    targetTLS.on('end', () => closeBothSockets());
+
+    this.internalTLS.on('error', () => closeBothSockets());
+    targetTLS.on('error', error => {
+      if (['DEPTH_ZERO_SELF_SIGNED_CERT', 'ERR_SSL_TLSV13_ALERT_CERTIFICATE_REQUIRED'].includes(error.code)) {
+        this.internalTLS!.write('HTTP/1.1 503 Internal Server Error\r\n');
```

Comment:
Could you please explain this scenario? I find it a bit strange that we handle some errors, but not others, so I'd like to understand it better.

Reply:
> Architecture: Browser -> ClientCertificatesProxy -> Target

When a TLS error happens between `ClientCertificatesProxy -> Target` we cannot give it back to the client how it actually should have get reported. So we need to wrap it inside a http response like above. We could and should probably do it for all TLS errors between proxy and target.